File: examples/selected_refinement_operator.cpp
#include <cstddef>
#include <cstdint>
#include <vector>

#include <tsl.hpp>

struct less_than_op {
    template <class Vec>
    typename Vec::mask_type operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        return tsl::less_than<Vec>(left, right);
    }
};

struct negative_op {
    template <class Vec>
    typename Vec::mask_type operator()(typename tsl::reg_param<Vec>::type value) const {
        return tsl::less_than<Vec>(
            value,
            tsl::set1<Vec>(static_cast<typename Vec::base_type>(0)));
    }
};

void fill_inputs(
    std::vector<std::int32_t>& left,
    std::vector<std::int32_t>& right) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        left[i] = static_cast<std::int32_t>(static_cast<int>((i * 19) % 67) - 33);
        right[i] = static_cast<std::int32_t>(static_cast<int>((i * 5) % 37) - 18);
    }
}

std::vector<std::uint32_t> make_selection(std::size_t count) {
    std::vector<std::uint32_t> indices;
    for (std::size_t i = count; i > 0; --i) {
        const std::size_t row = i - 1;
        if ((row % 2) == 0 || (row % 7) == 0) {
            indices.push_back(static_cast<std::uint32_t>(row));
        }
    }
    return indices;
}

template <class IndexT, class Predicate>
bool verify_refined_indices(
    const std::vector<std::uint32_t>& input_indices,
    const std::vector<IndexT>& output_indices,
    std::size_t produced,
    Predicate predicate) {
    constexpr IndexT sentinel = static_cast<IndexT>(999999);
    std::size_t expected_count = 0;
    for (const auto row : input_indices) {
        if (predicate(static_cast<std::size_t>(row))) {
            if (output_indices[expected_count] != static_cast<IndexT>(row)) {
                return false;
            }
            expected_count += 1;
        }
    }
    if (produced != expected_count) {
        return false;
    }
    for (std::size_t i = produced; i < output_indices.size(); ++i) {
        if (output_indices[i] != sentinel) {
            return false;
        }
    }
    return true;
}

template <std::size_t ParallelN>
bool run_selected_refinement_pointer_case() {
    constexpr std::size_t count = 1003;
    constexpr std::uint32_t sentinel = 999999;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    fill_inputs(left, right);
    const auto indices = make_selection(count);

    std::vector<std::uint32_t> refined(indices.size(), sentinel);
    auto produced = tsl::algo::select_selected_indices_unary<ParallelN>(
        negative_op{},
        left.data(),
        indices.data(),
        refined.data(),
        indices.size());
    if (!verify_refined_indices(
            indices,
            refined,
            produced,
            [&](std::size_t row) { return left[row] < 0; })) {
        return false;
    }

    refined.assign(indices.size(), sentinel);
    produced = tsl::algo::select_selected_indices_binary<ParallelN>(
        less_than_op{},
        left.data(),
        right.data(),
        indices.data(),
        refined.data(),
        indices.size());
    if (!verify_refined_indices(
            indices,
            refined,
            produced,
            [&](std::size_t row) { return left[row] < right[row]; })) {
        return false;
    }

    return true;
}

template <std::size_t ParallelN>
bool run_selected_refinement_range_case() {
    constexpr std::size_t count = 257;
    constexpr std::uint32_t sentinel = 999999;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    fill_inputs(left, right);
    const auto indices = make_selection(count);

    std::vector<std::uint32_t> refined(indices.size(), sentinel);
    const auto produced = tsl::algo::select_selected_indices_binary<ParallelN>(
        less_than_op{},
        left,
        right,
        indices,
        refined);
    return verify_refined_indices(
        indices,
        refined,
        produced,
        [&](std::size_t row) { return left[row] < right[row]; });
}

template <std::size_t ParallelN>
bool run_selected_refinement_cases() {
    return run_selected_refinement_pointer_case<ParallelN>() &&
           run_selected_refinement_range_case<ParallelN>();
}

int main() {
    if (!run_selected_refinement_cases<1>()) {
        return 1;
    }
    if (!run_selected_refinement_cases<4>()) {
        return 2;
    }
    if (!run_selected_refinement_cases<16>()) {
        return 3;
    }
    return 0;
}

File: examples/rust/src/bin/selected_refinement_operator.rs
use tsl::profile;
use tsl::tsl_core::StaticSimdVector;

struct LessThan;

impl<V> profile::algo::BinaryPredicateKernel<V> for LessThan
where
    V: StaticSimdVector<BaseType = i32> + profile::detail::primitives::Less_thanImpl,
{
    fn test(&mut self, left: V::RegisterType, right: V::RegisterType) -> V::MaskType {
        profile::less_than::<V>(left, right)
    }
}

struct Negative;

impl<V> profile::algo::UnaryPredicateKernel<V> for Negative
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::Less_thanImpl
        + profile::detail::primitives::Set1Impl,
{
    fn test(&mut self, value: V::RegisterType) -> V::MaskType {
        profile::less_than::<V>(value, profile::set1::<V>(0))
    }
}

fn fill_inputs(left: &mut [i32], right: &mut [i32]) {
    for (i, (left_value, right_value)) in left.iter_mut().zip(right.iter_mut()).enumerate() {
        *left_value = ((i * 19) % 67) as i32 - 33;
        *right_value = ((i * 5) % 37) as i32 - 18;
    }
}

fn make_selection(count: usize) -> Vec<usize> {
    let mut indices = Vec::new();
    for row in (0..count).rev() {
        if row % 2 == 0 || row % 7 == 0 {
            indices.push(row);
        }
    }
    indices
}

fn verify_refined_indices<F>(
    input_indices: &[usize],
    output_indices: &[usize],
    produced: usize,
    sentinel: usize,
    mut predicate: F,
) where
    F: FnMut(usize) -> bool,
{
    let mut expected_count = 0usize;
    for &row in input_indices {
        if predicate(row) {
            assert_eq!(output_indices[expected_count], row);
            expected_count += 1;
        }
    }
    assert_eq!(produced, expected_count);
    for value in output_indices.iter().skip(produced) {
        assert_eq!(*value, sentinel);
    }
}

fn main() {
    const COUNT: usize = 1003;
    const SENTINEL: usize = 999_999;

    let mut left = vec![0i32; COUNT];
    let mut right = vec![0i32; COUNT];
    fill_inputs(&mut left, &mut right);
    let indices = make_selection(COUNT);

    macro_rules! run_policy {
        ($policy:expr) => {{
            let policy = $policy;

            let mut refined = vec![SENTINEL; indices.len()];
            let mut negative = Negative;
            let produced = profile::algo::select_selected_indices_unary(
                policy,
                &mut negative,
                &left,
                &indices,
                &mut refined,
            );
            verify_refined_indices(&indices, &refined, produced, SENTINEL, |row| left[row] < 0);

            refined.fill(SENTINEL);
            let mut less_than = LessThan;
            let produced = profile::algo::select_selected_indices_binary(
                policy,
                &mut less_than,
                &left,
                &right,
                &indices,
                &mut refined,
            );
            verify_refined_indices(&indices, &refined, produced, SENTINEL, |row| {
                left[row] < right[row]
            });

            refined.fill(SENTINEL);
            let mut scaled_less_than = LessThan;
            let produced = unsafe {
                profile::algo::select_selected_indices_binary_scaled_raw::<4, _, _, _>(
                    policy,
                    &mut scaled_less_than,
                    left.as_ptr(),
                    right.as_ptr(),
                    indices.as_ptr(),
                    refined.as_mut_ptr(),
                    indices.len(),
                )
            };
            verify_refined_indices(&indices, &refined, produced, SENTINEL, |row| {
                left[row] < right[row]
            });
        }};
    }

    run_policy!(tsl::dataparallel::fixed::<1>());
    run_policy!(tsl::dataparallel::generic::<4>());
    run_policy!(tsl::dataparallel::generic::<16>());
}
